test/Tester/StorageFacet/Feature.Abstractions/IStorageFeature.cs

```diff
@@ -0,0 +1,58 @@
+﻿using System.Threading.Tasks;
+
+namespace Tester.StorageFacet.Abstractions
+{
+    /// <summary>
+    /// Primary storage feature interface.
+    ///  This is the actual functionality the users need.
+    /// </summary>
+    /// <typeparam name="TState"></typeparam>
+    public interface IStorageFeature<TState>
+    {
+        TState State { get; set; }
+
+        Task Save();
+
+        // Test calls - used to verify facet wiring works
+        string Name { get; }
+        string GetExtendedInfo();
+    }
+
+    /// <summary>
+    /// Feature configuration information which application layer can provide to the
+    ///  feature per instance (by grain type if using attributes).
+    /// </summary>
+    public interface IStorageFeatureConfig
+    {
+        string StateName { get; }
+    }
+
+    /// <summary>
+    /// Feature configuration utility class
+    /// </summary>
+    public class StorageFeatureConfig : IStorageFeatureConfig
+    {
+        public StorageFeatureConfig(string stateName)
+        {
+            this.StateName = stateName;
+        }
+
+        public string StateName { get; }
+    }
+
+    /// <summary>
+    /// Creates a storage feature from a configuration
+    /// </summary>
+    public interface IStorageFeatureFactory
+    {
+        IStorageFeature<TState> Create<TState>(IStorageFeatureConfig config);
```

Comment:
Instead of the non-generic interface with a generic method, could we have a generic interface? 
That would make the factory more explicit of what it can create. Sure, we can by default register the factory as an open generic, but that would allow us to override it for a specific state (explicit dependency principle).
Also, that would avoid a lot of MakeGenericMethod calls an each grain activation at runtime, and incur in that reflection cost only once per type (nevertheless this is an optimization, main reason is being explicit and allowing a different factory for a specific state type).